runtime/runtime/src/actions.rs

```diff
@@ -440,30 +441,39 @@ pub(crate) fn action_implicit_account_creation_transfer(
 ) {
     *actor_id = account_id.clone();
 
-    let mut access_key = AccessKey::full_access();
-    // Set default nonce for newly created access key to avoid transaction hash collision.
-    // See <https://github.com/near/nearcore/issues/3779>.
-    if checked_feature!("stable", AccessKeyNonceForImplicitAccounts, current_protocol_version) {
-        access_key.nonce = (block_height - 1)
-            * near_primitives::account::AccessKey::ACCESS_KEY_NONCE_RANGE_MULTIPLIER;
-    }
+    match account_id.get_account_type() {
+        AccountType::NearImplicitAccount => {
+            let mut access_key = AccessKey::full_access();
+            // Set default nonce for newly created access key to avoid transaction hash collision.
+            // See <https://github.com/near/nearcore/issues/3779>.
+            if checked_feature!(
+                "stable",
+                AccessKeyNonceForImplicitAccounts,
+                current_protocol_version
+            ) {
+                access_key.nonce = (block_height - 1)
+                    * near_primitives::account::AccessKey::ACCESS_KEY_NONCE_RANGE_MULTIPLIER;
+            }
 
-    // Invariant: The account_id is hex like (implicit account id).
-    // It holds because in the only calling site, we've checked the permissions before.
-    // unwrap: Can only fail if `account_id` is not implicit.
-    let public_key = PublicKey::from_implicit_account(account_id).unwrap();
+            // unwrap: here it's safe because the `account_id` has already been determined to be implicit by `get_account_type`
+            let public_key = PublicKey::from_near_implicit_account(account_id).unwrap();
 
-    *account = Some(Account::new(
-        transfer.deposit,
-        0,
-        CryptoHash::default(),
-        fee_config.storage_usage_config.num_bytes_account
-            + public_key.len() as u64
-            + borsh::object_length(&access_key).unwrap() as u64
-            + fee_config.storage_usage_config.num_extra_bytes_record,
-    ));
+            *account = Some(Account::new(
+                transfer.deposit,
+                0,
+                CryptoHash::default(),
+                fee_config.storage_usage_config.num_bytes_account
+                    + public_key.len() as u64
+                    + borsh::object_length(&access_key).unwrap() as u64
+                    + fee_config.storage_usage_config.num_extra_bytes_record,
+            ));
 
-    set_access_key(state_update, account_id.clone(), public_key, &access_key);
+            set_access_key(state_update, account_id.clone(), public_key, &access_key);
+        }
+        // Invariant: The `account_id` is implicit.
+        // It holds because in the only calling site, we've checked the permissions before.
```

Comment:
sanity check - did you check that this is the case?

Reply:
Yes, it is checked in https://github.com/near/nearcore/blob/4319933bc807c306fa300cdc4251d0ae3ea5e306/runtime/runtime/src/actions.rs#L919